sql: add ADD REGION IF NOT EXISTS / DROP REGION IF EXISTS

otan opened the pull request:
See individual commits for details.

I'd be up for backporting this, it simplifies idempotent test setups for me a lot!

otan committed:
sql: fix error message for dropping non-added region

Release note (sql change): Previously, the error message for not having
a region defined on a database resulted in an error about enums. This is
now clearer that it is an undefined region.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/65641)
<!-- Reviewable:end -->

awoods187 commented:
I support backporting this from a product perspective. Nicely done.

On Mon, May 24, 2021 at 6:29 PM cockroach-teamcity ***@***.***>
wrote:

> This change is [image: Reviewable]
> <https://reviewable.io/reviews/cockroachdb/cockroach/65641>
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/65641#issuecomment-847392681>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFJ7F7Y3SCWXVLK5ZD3KIQDTPLHLVANCNFSM45OBZGQQ>
> .
>


-- 
Andy Woods
Group Product Manager, Cockroach Labs
***@***.***

otan committed:
sql: add ALTER DATABASE ... ADD REGION IF NOT EXISTS ...

Found I wanted this multiple times already for idempotency, so added the
syntax.

Release note (sql change): Implement ALTER DATABASE ... ADD REGION IF
NOT EXISTS ... which does not error if adding a region that is already
in the database.

otan committed:
sql: add ALTER DATABASE ... DROP REGION IF EXISTS ...

Release note (sql change): Add ALTER DATABASE ... DROP REGION IF EXISTS
... syntax, which does not error if dropping a region that is not
defined on the database.

arulajmani reviewed (commented):
Let's buffer a notice in the case the operation is a no-op? We seem to be doing that with a few other "IF EXISTS" style statements lately. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65641#-:-MaWBszKAfsDcxStpopp:b-r2f16s)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm)



<!-- Sent from Reviewable.io -->

otan commented:
> Let's buffer a notice in the case the operation is a no-op? We seem to be doing that with a few other "IF EXISTS" style statements lately.

really? seems very unpostgres-like :P

arulajmani reviewed (commented):
#62073 as an example, with the postgres example snippet -- shouldn't we be consistent?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65641#-:-MaWFfbu9Xvi9OICtLBR:bqwo964)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm)



<!-- Sent from Reviewable.io -->

otan commented:
haha, that's new!

ajstorm reviewed (commented):
> Reviewed 5 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65641)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/sql/parser/sql.y, line 1580 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65641#-MaYmUU78Wr1dDYngygi:-MaYmUU78Wr1dDYngygj:b-4cmth7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6384eb28b60450682952ffcbbe271a81396daa8/pkg/sql/parser/sql.y#L1580)):*
> ```Yacc
>     }
>   }
> | ALTER DATABASE database_name ADD REGION IF NOT EXISTS region_name
> ```

Nit: Does it make sense to move the IF NOT EXISTS into a separate opt clause so that we can keep a single `ALTER DATABASE...ADD REGION` block?  I'd think that would make this code more maintainable (less room for error when updating things, but I see that the precedent in sql.y is to have multiple blocks when there's an IF NOT EXISTS.

i.e.

`ALTER DATABASE database_name ADD REGION opt_if_not_exists region_name`


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/parser/sql.y, line 1580 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65641#-MaYmUU78Wr1dDYngygi:-Ma_8B7--STeQBKzx61F:b-99skgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6384eb28b60450682952ffcbbe271a81396daa8/pkg/sql/parser/sql.y#L1580)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: Does it make sense to move the IF NOT EXISTS into a separate opt clause so that we can keep a single `ALTER DATABASE...ADD REGION` block?  I'd think that would make this code more maintainable (less room for error when updating things, but I see that the precedent in sql.y is to have multiple blocks when there's an IF NOT EXISTS.

i.e.

`ALTER DATABASE database_name ADD REGION opt_if_not_exists region_name`
</blockquote></details>

you have the same idea as i had a year ago :)

shift-reduce errors from YACC will stop you.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/parser/sql.y, line 1580 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65641#-MaYmUU78Wr1dDYngygi:-Ma_8I778zmL8lZ1nZbw:b-kiddxu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6384eb28b60450682952ffcbbe271a81396daa8/pkg/sql/parser/sql.y#L1580)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

you have the same idea as i had a year ago :)

shift-reduce errors from YACC will stop you.
</blockquote></details>

(that's why we don't do it anywhere)


<!-- Sent from Reviewable.io -->

ajstorm reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65641)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/sql/parser/sql.y, line 1580 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65641#-MaYmUU78Wr1dDYngygi:-Ma_UJ_G3cTM4h7o8_t3:b-1uwq5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6384eb28b60450682952ffcbbe271a81396daa8/pkg/sql/parser/sql.y#L1580)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

(that's why we don't do it anywhere)
</blockquote></details>

I'm always (at least) one year behind you.

Do you recall what specific shift-reduce errors you got?


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/parser/sql.y, line 1580 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65641#-MaYmUU78Wr1dDYngygi:-Ma_W46z7QTq2z8ZpvPu:bploz77) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6384eb28b60450682952ffcbbe271a81396daa8/pkg/sql/parser/sql.y#L1580)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

I'm always (at least) one year behind you.

Do you recall what specific shift-reduce errors you got?
</blockquote></details>

that's the only thing you see when there's an error!


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65641)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm)

___
*[pkg/sql/parser/sql.y, line 1580 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65641#-MaYmUU78Wr1dDYngygi:-Ma_aZ9P-LlprhdKFVKi:b-ocnpfp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6384eb28b60450682952ffcbbe271a81396daa8/pkg/sql/parser/sql.y#L1580)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

that's the only thing you see when there's an error!
</blockquote></details>

```
 ⚙ otan@otan-mbp  ~/go/src/github.com/cockroachdb/cockroach   if_exists ●  make buildshort
Running make with -j16
GOPATH set to /Users/otan/go
mkdir -p lib
set -euo pipefail; \
  ret=$(cd pkg/sql/parser/gen && goyacc -p sql -o sql.go.tmp sql-gen.y); \
  if expr "$ret" : ".*conflicts" >/dev/null; then \
    echo "$ret"; exit 1; \
  fi
git diff -r rm -f lib/lib{geos,geos_c}.dylib
cp -L /Users/otan/go/native/x86_64-apple-darwin19.6.0/geos/lib/lib{geos,geos_c}.dylib lib
HEAD
conflicts: 1 shift/reduce
gmake: *** [Makefile:1451: pkg/sql/parser/gen/sql.go.tmp] Error 1
gmake: *** Deleting file 'pkg/sql/parser/gen/sql.go.tmp'
 ✘ ⚙ otan@otan-mbp  ~/go/src/github.com/cockroachdb/cockroach   if_exists ●  git diff -r HEAD
diff --git a/pkg/sql/parser/sql.y b/pkg/sql/parser/sql.y
index eb4feca835..43dfe59f88 100644
--- a/pkg/sql/parser/sql.y
+++ b/pkg/sql/parser/sql.y
@@ -1091,6 +1091,7 @@ func (u *sqlSymUnion) objectNamePrefixList() tree.ObjectNamePrefixList {
 %type <empty> opt_all_clause
 %type <empty> opt_privileges_clause
 %type <bool> distinct_clause
+%type <bool> opt_if_not_exists
 %type <tree.DistinctOn> distinct_on_clause
 %type <tree.NameList> opt_column_list insert_column_list opt_stats_columns
 %type <tree.OrderBy> sort_clause single_sort_clause opt_sort_clause
@@ -1570,19 +1571,12 @@ alter_database_owner:
   }

 alter_database_add_region_stmt:
-  ALTER DATABASE database_name ADD REGION region_name
+  ALTER DATABASE database_name ADD REGION opt_if_not_exists region_name
   {
     $$.val = &tree.AlterDatabaseAddRegion{
       Name: tree.Name($3),
-      Region: tree.Name($6),
-    }
-  }
-| ALTER DATABASE database_name ADD REGION IF NOT EXISTS region_name
-  {
-    $$.val = &tree.AlterDatabaseAddRegion{
-      Name: tree.Name($3),
-      Region: tree.Name($9),
-      IfNotExists: true,
+      Region: tree.Name($7),
+      IfNotExists: $6.bool(),
     }
   }

@@ -8857,6 +8851,16 @@ all_or_distinct:
     $$.val = false
   }

+opt_if_not_exists:
+  IF NOT EXISTS
+  {
+    $$.val = true
+  }
+| /* EMPTY */
+  {
+    $$.val = false
+  }
+
 distinct_clause:
   DISTINCT
   {
```


<!-- Sent from Reviewable.io -->

ajstorm reviewed (approved):
> Reviewed 1 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65641)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/sql/parser/sql.y, line 1580 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65641#-MaYmUU78Wr1dDYngygi:-Macsq0wBP6wiqMEjsC0:b-q7iq1r) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6384eb28b60450682952ffcbbe271a81396daa8/pkg/sql/parser/sql.y#L1580)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

```
 ⚙ otan@otan-mbp  ~/go/src/github.com/cockroachdb/cockroach   if_exists ●  make buildshort
Running make with -j16
GOPATH set to /Users/otan/go
mkdir -p lib
set -euo pipefail; \
  ret=$(cd pkg/sql/parser/gen && goyacc -p sql -o sql.go.tmp sql-gen.y); \
  if expr "$ret" : ".*conflicts" >/dev/null; then \
    echo "$ret"; exit 1; \
  fi
git diff -r rm -f lib/lib{geos,geos_c}.dylib
cp -L /Users/otan/go/native/x86_64-apple-darwin19.6.0/geos/lib/lib{geos,geos_c}.dylib lib
HEAD
conflicts: 1 shift/reduce
gmake: *** [Makefile:1451: pkg/sql/parser/gen/sql.go.tmp] Error 1
gmake: *** Deleting file 'pkg/sql/parser/gen/sql.go.tmp'
 ✘ ⚙ otan@otan-mbp  ~/go/src/github.com/cockroachdb/cockroach   if_exists ●  git diff -r HEAD
diff --git a/pkg/sql/parser/sql.y b/pkg/sql/parser/sql.y
index eb4feca835..43dfe59f88 100644
--- a/pkg/sql/parser/sql.y
+++ b/pkg/sql/parser/sql.y
@@ -1091,6 +1091,7 @@ func (u *sqlSymUnion) objectNamePrefixList() tree.ObjectNamePrefixList {
 %type <empty> opt_all_clause
 %type <empty> opt_privileges_clause
 %type <bool> distinct_clause
+%type <bool> opt_if_not_exists
 %type <tree.DistinctOn> distinct_on_clause
 %type <tree.NameList> opt_column_list insert_column_list opt_stats_columns
 %type <tree.OrderBy> sort_clause single_sort_clause opt_sort_clause
@@ -1570,19 +1571,12 @@ alter_database_owner:
   }

 alter_database_add_region_stmt:
-  ALTER DATABASE database_name ADD REGION region_name
+  ALTER DATABASE database_name ADD REGION opt_if_not_exists region_name
   {
     $$.val = &tree.AlterDatabaseAddRegion{
       Name: tree.Name($3),
-      Region: tree.Name($6),
-    }
-  }
-| ALTER DATABASE database_name ADD REGION IF NOT EXISTS region_name
-  {
-    $$.val = &tree.AlterDatabaseAddRegion{
-      Name: tree.Name($3),
-      Region: tree.Name($9),
-      IfNotExists: true,
+      Region: tree.Name($7),
+      IfNotExists: $6.bool(),
     }
   }

@@ -8857,6 +8851,16 @@ all_or_distinct:
     $$.val = false
   }

+opt_if_not_exists:
+  IF NOT EXISTS
+  {
+    $$.val = true
+  }
+| /* EMPTY */
+  {
+    $$.val = false
+  }
+
 distinct_clause:
   DISTINCT
   {
```
</blockquote></details>

Thanks for the added context.  Change LGTM.  


<!-- Sent from Reviewable.io -->

otan commented:
bors r=ajstorm
